File: Library/src/com/androidanimator/animation/RotationAnimation.java
package com.androidanimator.animation;

import com.androidanimator.animation.AnimationListener;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.view.View;
import android.view.ViewGroup;

/**
 * 
 * @author SiYao
 *
 */
public class RotationAnimation extends Animation {

	public static final int PIVOT_CENTER = 0, PIVOT_TOP_LEFT = 1,
			PIVOT_TOP_RIGHT = 2, PIVOT_BOTTOM_LEFT = 3, PIVOT_BOTTOM_RIGHT = 4;

	float degrees;
	int pivot;
	long duration;
	AnimationListener listener;

	/**
	 * The RotationAnimation rotates the view depending on the parameters
	 * provided by the user.
	 * 
	 * @param degrees
	 *            the number of degrees to rotate the view by
	 * @param pivot
	 *            the pivot around which the view is rotated
	 * @param duration
	 *            the duration of the entire animation
	 * @param listener
	 *            the AnimationListener of animation @see
	 *            {@link AnimationListener}
	 */
	public RotationAnimation(View view) {
		this.view = view;
		degrees = 360;
		pivot = PIVOT_CENTER;
		duration = Animation.DEFAULT_DURATION;
		listener = null;
	}
	
	@Override
	public void animate() {
		ViewGroup parentView = (ViewGroup) view.getParent(), rootView = (ViewGroup) view
				.getRootView();
		while (parentView != rootView) {
			parentView.setClipChildren(false);
			parentView = (ViewGroup) parentView.getParent();
		}
		rootView.setClipChildren(false);

		float pivotX, pivotY, viewWidth = view.getWidth(), viewHeight = view.getHeight();
		switch (pivot) {
		case PIVOT_TOP_LEFT:
			pivotX = 1f;
			pivotY = 1f;
			break;
		case PIVOT_TOP_RIGHT:
			pivotX = viewWidth;
			pivotY = 1f;
			break;
		case PIVOT_BOTTOM_LEFT:
			pivotX = 1f;
			pivotY = viewHeight;
			break;
		case PIVOT_BOTTOM_RIGHT:
			pivotX = viewWidth;
			pivotY = viewHeight;
			break;
		default:
			pivotX = viewWidth/2;
			pivotY = viewHeight/2;
			break;
		}
		view.setPivotX(pivotX);
		view.setPivotY(pivotY);
		view.animate().rotationBy(degrees).setDuration(duration).setListener(new AnimatorListenerAdapter() {

			@Override
			public void onAnimationEnd(Animator animation) {
				if (getListener() != null) {
					getListener().onAnimationEnd(RotationAnimation.this);
				}
			}
		});
	}

	/**
	 * @return the degrees
	 */
	public float getDegrees() {
		return degrees;
	}

	/**
	 * @param degrees the degrees to set
	 */
	public RotationAnimation setDegrees(float degrees) {
		this.degrees = degrees;
		return this;
	}

	/**
	 * @return the pivot
	 */
	public int getPivot() {
		return pivot;
	}

	/**
	 * @param pivot the pivot to set
	 */
	public RotationAnimation setPivot(int pivot) {
		this.pivot = pivot;
		return this;
	}

	/**
	 * @return the duration
	 */
	public long getDuration() {
		return duration;
	}

	/**
	 * @param duration the duration to set
	 */
	public RotationAnimation setDuration(long duration) {
		this.duration = duration;
		return this;
	}

	/**
	 * @return the listener
	 */
	public AnimationListener getListener() {
		return listener;
	}

	/**
	 * @param listener the listener to set
	 */
	public RotationAnimation setListener(AnimationListener listener) {
		this.listener = listener;
		return this;
	}

}

File: Library/src/com/easyandroidanimations/library/kotlinlib/RotationAnimation.kt
package com.easyandroidanimations.library

import android.animation.Animator

/**
 * This animation rotates the view by a customizable number of degrees and at a
 * customizable pivot point.
 *
 * @author SiYao
 */
class RotationAnimation(view: View) : Animation(), Combinable {
    /**
     * @return The number of degrees to rotate by.
     */
    var degrees: Float

    /**
     * The available pivot points are `PIVOT_CENTER`,
     * `PIVOT_TOP_LEFT`, `PIVOT_TOP_RIGHT`,
     * `PIVOT_BOTTOM_LEFT` and `PIVOT_BOTTOM_RIGHT`.
     *
     * @return The pivot point for rotation.
     */
    var pivot: Int
    var interpolator: TimeInterpolator

    /**
     * @return The duration of the entire animation.
     */
    override var duration: Long
    var listener: AnimationListener?

    @Override
    override fun animate() {
        animatorSet.start()
    }

    @get:Override
    override val animatorSet: AnimatorSet
        get() {
            var parentView: ViewGroup = view.getParent() as ViewGroup
            val rootView: ViewGroup = view.getRootView() as ViewGroup
            while (parentView !== rootView) {
                parentView.setClipChildren(false)
                parentView = parentView.getParent() as ViewGroup
            }
            rootView.setClipChildren(false)
            val pivotX: Float
            val pivotY: Float
            val viewWidth: Float = view.getWidth()
            val viewHeight: Float = view.getHeight()
            when (pivot) {
                PIVOT_TOP_LEFT -> {
                    pivotX = 1f
                    pivotY = 1f
                }
                PIVOT_TOP_RIGHT -> {
                    pivotX = viewWidth
                    pivotY = 1f
                }
                PIVOT_BOTTOM_LEFT -> {
                    pivotX = 1f
                    pivotY = viewHeight
                }
                PIVOT_BOTTOM_RIGHT -> {
                    pivotX = viewWidth
                    pivotY = viewHeight
                }
                else -> {
                    pivotX = viewWidth / 2
                    pivotY = viewHeight / 2
                }
            }
            view.setPivotX(pivotX)
            view.setPivotY(pivotY)
            val rotationSet = AnimatorSet()
            rotationSet.play(ObjectAnimator.ofFloat(view, View.ROTATION, view.getRotation() + degrees))
            rotationSet.setInterpolator(interpolator)
            rotationSet.setDuration(duration)
            rotationSet.addListener(object : AnimatorListenerAdapter() {
                @Override
                fun onAnimationEnd(animation: Animator?) {
                    if (getListener() != null) {
                        getListener().onAnimationEnd(this@RotationAnimation)
                    }
                }
            })
            return rotationSet
        }

    /**
     * In order to rotate anti-clockwise, the number of degrees should be
     * negative.
     *
     * @param degrees
     * The number of degrees to set to rotate by.
     * @return This object, allowing calls to methods in this class to be
     * chained.
     */
    fun setDegrees(degrees: Float): RotationAnimation {
        this.degrees = degrees
        return this
    }

    /**
     * The available pivot points are `PIVOT_CENTER`,
     * `PIVOT_TOP_LEFT`, `PIVOT_TOP_RIGHT`,
     * `PIVOT_BOTTOM_LEFT` and `PIVOT_BOTTOM_RIGHT`.
     *
     * @param pivot
     * The pivot point to set for rotation.
     * @return This object, allowing calls to methods in this class to be
     * chained.
     */
    fun setPivot(pivot: Int): RotationAnimation {
        this.pivot = pivot
        return this
    }

    /**
     * @return The interpolator of the entire animation.
     */
    fun getInterpolator(): TimeInterpolator {
        return interpolator
    }

    /**
     * @param interpolator
     * The interpolator of the entire animation to set.
     */
    override fun setInterpolator(interpolator: TimeInterpolator): RotationAnimation {
        this.interpolator = interpolator
        return this
    }

    /**
     * @param duration
     * The duration of the entire animation to set.
     * @return This object, allowing calls to methods in this class to be
     * chained.
     */
    override fun setDuration(duration: Long): RotationAnimation {
        this.duration = duration
        return this
    }

    /**
     * @return The listener for the end of the animation.
     */
    fun getListener(): AnimationListener? {
        return listener
    }

    /**
     * @param listener
     * The listener to set for the end of the animation.
     * @return This object, allowing calls to methods in this class to be
     * chained.
     */
    override fun setListener(listener: AnimationListener?): RotationAnimation {
        this.listener = listener
        return this
    }

    companion object {
        const val PIVOT_CENTER = 0
        const val PIVOT_TOP_LEFT = 1
        const val PIVOT_TOP_RIGHT = 2
        const val PIVOT_BOTTOM_LEFT = 3
        const val PIVOT_BOTTOM_RIGHT = 4
    }

    /**
     * This animation rotates the view by a customizable number of degrees and
     * at a customizable pivot point.
     *
     * @param view
     * The view to be animated.
     */
    init {
        view = view
        degrees = 360f
        pivot = PIVOT_CENTER
        interpolator = AccelerateDecelerateInterpolator()
        duration = DURATION_LONG
        listener = null
    }
}